scanner/decisions/0001-scannerv4-apis.md

```diff
@@ -0,0 +1,93 @@
+
+### ScannerV4 API Endpoints
+
+The endpoints are versioned at `v4` to align with Clair:
+
+1. `scanner.v4.Indexer/CreateIndex`: Create a manifest of a specified resource and create an index or re-index. Idempotent while creation is being executed. Synchronous and tied to the client's request. Returns [`IndexReport`](https://github.com/quay/claircore/blob/v1.4.18/indexreport.go#L19)
+2. `scanner.v4.Indexer/GetIndex`: Retrieve or check index existence. Returns [`IndexReport`](https://github.com/quay/claircore/blob/v1.4.18/indexreport.go#L19).
+3. `scanner.v4.Matcher/GetVulnerabilities`: Get vulnerabilities for a given resource's manifest. Returns [`VulnerabilityReport`](https://github.com/quay/claircore/blob/v1.4.18/vulnerabilityreport.go#L7).
+4. `scanner.v4.Matcher/GetVulnerabilityMetadata`: Get information on vulnerability metadata, e.g., last update timestamp.
+
+Example of manifest's `hash_id` usage to create index reports for container images:
+
+```
+message ContainerImageLocator {
+    string url      = 1;
+    string username = 2;
+    string password = 3;
+}
+
+message CreateIndexRequest {
+    string hash_id = 1;
+    oneof resource_locator {
+        ContainerImageLocator container_image;
+    }
+}
+```
+
+Example of index request and vulnerability request:
+
+```
+message GetIndexRequest {
+    string hash_id   = 1;
+    // If true, does not return the Index Report, only status.
+    bool check_only  = 2;
+}
+
+message GetVulnerabilitiesRequest {
+    string hash_id = 1;
+}
+```
+
+Examples of common error codes and scenarios:
+
+| Code | Scenario |
+| --- | ---
+| `UNAVAILABLE` | Scanner temporary or retriable issue caused by dependencies (e.g., registries, databases).
+| `INVALID_ARGUMENT` | Invalid input or malformed input (e.g., bad credentials, malformed `hash_id`).
+| `INTERNAL` | Unrecoverable errors, unexpected conditions, or invariant (e.g., bugs, broken schema, etc).
+| `FAILED_PRECONDITION	` | Index report does not exist for this given `hash_id` (e.g. when calling `GetVulnerabilities`).
+
+All APIs are idempotent, hence retriable on timeouts and temporary errors.
+
+## Consequences
+
+Leveraging gRPC-only helps maintain service contracts and enhances performance. It also aligns with other Stackrox services.
+
+The new image integration for ScannerV4 ensures minimal changes are necessary for Central to switch between ScannerV2 and ScannerV4 using a feature flag.
+
+The Indexer and Matcher modes allow additional deployment options, enabling ScannerV4 to scale and serve different scenarios. The multiple modes and gRPC services pave the way for advanced deployments, such as multi-tenant scanning and node scanning.
+
+Using gRPC exclusively increases complexity for load balancing ScannerV4 for horizontal scalability. Additional steps will be necessary to achieve this, including service-mesh tooling or client-side load-balancing, which are not available out of the box and require additional work.
```

Comment:
This was a real killer when choosing to just use http+json in clair instead of grpc.

Reply:
We didn't do a deep analysis -- we were hitting response-size limits and then when we realized that the load balancing would be much worse, we just went with http+json.

Reply:
With gRPC + client-side round-robin, how would it keep the list of endpoints up-to-date, in a scenario where you are deploying or scaling a scanner?

I think that with a Kubernetes service, AFAIK, that would be handled by the cluster because the Kubernetes service knows the pod's lifecycle. 

If we introduce RESTful HTTP endpoints as a fallback for unreliable client-side load balancing, we might make it the standard way of calling Scanner in Central. We lose some of the benefits of HTTP/2, but still benefit from the proto and gRPC wiring.

Reply:
@jvdm:

> It's a pain.